File: back/node/app/delivery/middlewares/session/Session.ts
import session from 'express-session';
import RedisStore from 'connect-redis';
import { SESSION_SECRET } from '../../../connections/config/environment.js';
import {RedisClientType} from "redis";

const createSessionMiddleware = (redisClient: RedisClientType) => {
    const store: RedisStore = new RedisStore({
        client: redisClient,
        prefix: "tictac:",
    });

    const sessionSecret = SESSION_SECRET ? SESSION_SECRET : "secret"

    return session({
        name: "tictacnode",
        store: store,
        secret: sessionSecret,
        resave: false,
        saveUninitialized: false,
        cookie: {
            domain: 'localhost',
            path: '/',
            maxAge: 1000 * 60 * 60 * 24 * 30, // 30 days
            secure: false,
            httpOnly: true,
            sameSite: 'none',
        }
    });
};

export {createSessionMiddleware}

File: back/go/app/delivery/middlewares/session/Session.go
package middlewares

import(
	"fmt"
	"github.com/gorilla/sessions"
	"net/http"
)

func SessionMiddleware(store *sessions.CookieStore) func(next http.Handler) http.Handler {

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			session, err := store.Get(r, "tictacgo")
			if err != nil {
				http.Error(w, err.Error(), http.StatusInternalServerError)
				return
			}

			session.Options = &sessions.Options{
				Domain:   "localhost",
				Path:     "/",
				MaxAge:   30 * 24 * 60 * 60,
				Secure:   false,
				HttpOnly: true,
				SameSite: http.SameSiteLaxMode,
			}

			next.ServeHTTP(w, r)

			err = session.Save(r, w)
			if err != nil {
				fmt.Println("Error when trying to save session")
				fmt.Println(err.Error())
				return
			}
		})
	}
}

